File: src/repositories/CooldownRepository.js
/**
 * Cooldown Repository
 * Handles command cooldown tracking
 */

const { BaseRepository } = require('./BaseRepository');

/**
 * Repository for cooldowns table
 */
class CooldownRepository extends BaseRepository {
  /**
   * Create CooldownRepository instance
   * @param {Object} pool - PostgreSQL pool
   */
  constructor(pool) {
    super(pool, 'cooldowns', 'id');
  }

  /**
   * Get active cooldown for user and command
   * @param {string} userId - User ID
   * @param {string} command - Command name
   * @returns {Promise<Date|null>} Expiration time or null
   */
  async getCooldown(userId, command) {
    const rows = await this.findWhere(
      { userId, command },
      { orderBy: 'expires_at', order: 'DESC', limit: 1 }
    );

    if (rows.length === 0) {
      return null;
    }

    const expiresAt = new Date(rows[0].expires_at);

    // Check if expired
    if (expiresAt <= new Date()) {
      return null;
    }

    return expiresAt;
  }

  /**
   * Set cooldown for user and command
   * @param {string} userId - User ID
   * @param {string} command - Command name
   * @param {number} durationMs - Duration in milliseconds
   * @returns {Promise<boolean>}
   */
  async setCooldown(userId, command, durationMs) {
    const expiresAt = new Date(Date.now() + durationMs);

    try {
      await this.query(
        `
          INSERT INTO ${this.tableName} (user_id, command, expires_at, created_at)
          VALUES ($1, $2, $3, CURRENT_TIMESTAMP)
          ON CONFLICT (user_id, command)
          DO UPDATE SET expires_at = $3, created_at = CURRENT_TIMESTAMP
        `,
        [userId, command, expiresAt]
      );
      return true;
    } catch (error) {
      this.logger.error(`Failed to set cooldown: ${error.message}`);
      return false;
    }
  }

  /**
   * Check if user is on cooldown
   * @param {string} userId - User ID
   * @param {string} command - Command name
   * @returns {Promise<boolean>}
   */
  async isOnCooldown(userId, command) {
    const expiresAt = await this.getCooldown(userId, command);
    return expiresAt !== null;
  }

  /**
   * Get remaining cooldown time in milliseconds
   * @param {string} userId - User ID
   * @param {string} command - Command name
   * @returns {Promise<number>} Remaining ms (0 if not on cooldown)
   */
  async getRemainingTime(userId, command) {
    const expiresAt = await this.getCooldown(userId, command);

    if (!expiresAt) {
      return 0;
    }

    const remaining = expiresAt.getTime() - Date.now();
    return Math.max(0, remaining);
  }

  /**
   * Clear cooldown for user and command
   * @param {string} userId - User ID
   * @param {string} command - Command name
   * @returns {Promise<boolean>}
   */
  async clearCooldown(userId, command) {
    const deleted = await this.deleteWhere({ userId, command });
    return deleted > 0;
  }

  /**
   * Clear all expired cooldowns
   * @returns {Promise<number>} Number of cleared cooldowns
   */
  async clearExpired() {
    const result = await this.query(
      `DELETE FROM ${this.tableName} WHERE expires_at < NOW()`
    );

    return result?.rowCount || 0;
  }

  /**
   * Clear all cooldowns for a user
   * @param {string} userId - User ID
   * @returns {Promise<number>} Number of cleared cooldowns
   */
  async clearUserCooldowns(userId) {
    return this.deleteWhere({ userId });
  }
}

module.exports = { CooldownRepository };

File: repositories/cooldown_repository.py
"""
Cooldown Repository
Handles command cooldown tracking
"""

from datetime import datetime
from typing import Optional
import asyncpg

from repositories.base_repository import BaseRepository


class CooldownRepository(BaseRepository):
    """Repository for cooldowns table."""

    def __init__(self, pool: asyncpg.Pool):
        """
        Create CooldownRepository instance.

        Args:
            pool: PostgreSQL connection pool
        """
        super().__init__(pool, "cooldowns", "id")

    async def get_cooldown(self, user_id: str, command: str) -> Optional[datetime]:
        """
        Get active cooldown for user and command.

        Args:
            user_id: User ID
            command: Command name

        Returns:
            Expiration datetime or None
        """
        rows = await self.find_where(
            {"user_id": user_id, "command": command},
            {"order_by": "expires_at", "order": "DESC", "limit": 1}
        )

        if not rows:
            return None

        expires_at = rows[0].get("expires_at")
        if not expires_at:
            return None

        # Handle both datetime objects and strings
        if isinstance(expires_at, str):
            expires_at = datetime.fromisoformat(expires_at)

        # Check if expired
        if expires_at <= datetime.now():
            return None

        return expires_at

    async def set_cooldown(self, user_id: str, command: str, duration_ms: int) -> bool:
        """
        Set cooldown for user and command.

        Args:
            user_id: User ID
            command: Command name
            duration_ms: Duration in milliseconds

        Returns:
            True if successful
        """
        expires_at = datetime.now().timestamp() + (duration_ms / 1000)
        expires_datetime = datetime.fromtimestamp(expires_at)

        sql = f"""
            INSERT INTO {self.table_name} (user_id, command, expires_at, created_at)
            VALUES ($1, $2, $3, CURRENT_TIMESTAMP)
            ON CONFLICT (user_id, command)
            DO UPDATE SET expires_at = $3, created_at = CURRENT_TIMESTAMP
        """

        try:
            await self.query(sql, [user_id, command, expires_datetime])
            return True
        except Exception as e:
            self.logger.error(f"Failed to set cooldown: {e}")
            return False

    async def is_on_cooldown(self, user_id: str, command: str) -> bool:
        """
        Check if user is on cooldown.

        Args:
            user_id: User ID
            command: Command name

        Returns:
            True if on cooldown
        """
        expires_at = await self.get_cooldown(user_id, command)
        return expires_at is not None

    async def get_remaining_time(self, user_id: str, command: str) -> int:
        """
        Get remaining cooldown time in milliseconds.

        Args:
            user_id: User ID
            command: Command name

        Returns:
            Remaining milliseconds (0 if not on cooldown)
        """
        expires_at = await self.get_cooldown(user_id, command)

        if not expires_at:
            return 0

        remaining = expires_at.timestamp() - datetime.now().timestamp()
        return max(0, int(remaining * 1000))

    async def clear_cooldown(self, user_id: str, command: str) -> bool:
        """
        Clear cooldown for user and command.

        Args:
            user_id: User ID
            command: Command name

        Returns:
            True if cleared
        """
        deleted = await self.delete_where({"user_id": user_id, "command": command})
        return deleted > 0

    async def clear_expired(self) -> int:
        """
        Clear all expired cooldowns.

        Returns:
            Number of cleared cooldowns
        """
        sql = f"DELETE FROM {self.table_name} WHERE expires_at < NOW()"

        if not self.is_connected():
            self.logger.warning("Database not connected, clear_expired skipped")
            return 0

        try:
            async with self.pool.acquire() as conn:
                result = await conn.execute(sql)
                return int(result.split()[1]) if result and "DELETE" in result else 0
        except Exception as e:
            self.logger.error(f"Failed to clear expired cooldowns: {e}")
            raise

    async def clear_user_cooldowns(self, user_id: str) -> int:
        """
        Clear all cooldowns for a user.

        Args:
            user_id: User ID

        Returns:
            Number of cleared cooldowns
        """
        return await self.delete_where({"user_id": user_id})
